### ISSUE TITLE
RFC: removing numpy and pytorch and going towards a jax-only world

### ISSUE THREAD

**Issue by PhilipVinc** [netket/netket#480]
This discussion refers to v3.0. I'd appreciate comments from _everyone_ (seriously, everyone).

Currently the un-shipped v3.0 has 3 backends, that is, 3 frameworks you can use to define your neural networks:
 - numpy, there mainly for historical reasons, where you have to hand write everything in numpy (and numba to get decent performance). Of course you also have to hand-code the derivative of your network.
 - Jax, which has seen a lot of recent developments. It's usually at least if not more efficient than numpy/numba. Defining networks is easy, and you get the derivative for free. With my recent PR we now also support MPI operations for SR, but we don't have as many SR solvers as for numpy (diagonalization for example). Very soon this will also allow to support multiGPU execution without changing any code.
 - PyTorch, which is similar to jax in that networks are easy to define and automatically computes the gradient. However SR is very slow due to limitations in PyTorch which won't be addressed soon by developers. You can't use MPI with PyTorch unless someone writes the equivalent of mpi4jax for PyTorch, something that I'm not gonna do and requires a lot of effort.

Me and @gcarleo have recently been discussing about simplifying the internals of netket notably to fully exploit Automatic Differentiation in order to write more compact drivers for unsupervised/supervised learning and other future applications.

Simplyifying the internals mean, mainly, removing back-ends. Ideally all but one. Our current idea is that numpy can be dropped as it's quite unflexible, provided we port to Jax/PyTorch the missing features.
However, we have also been toying with the idea of removing PyTorch support, in order to have a much more compact, functional and streamlined codebase.

This would also simplify the API, because we wouldn't need to discriminate between the backends anymore.

What do people think? Can anyone defend strongly PyTorch and give a valid reason for keeping it around? How many people are intensely using it?

cc @femtobit @kchoo1118 @twesterhout @fabienalet and those who contributed the PyTorch backend @ChenAo-Phys @nikita-astronaut

**Comment by gcarleo** [comment]
Yes, let me add that implementing some applications (supervised, unsupervised, circuit learning etc) with numpy is quite painful, so it is highly likely that if we want to have some of these applications in 3.0 these will work only with JAX. 

Personally, I am totally **in favour of removing the Numpy backend** (i.e. mostly the machines and maybe the samplers) at this stage. Essentially, the only missing feature would be implementing RbmSpinSym in JAX, that should be possible to do relatively easily....

**Comment by femtobit** [comment]
@PhilipVinc Thanks for bringing this up. I very much agree with your points. Focusing on one backend would definitely make development easier. The fact that we can now cleanly integrate MPI operations with JAX is also a big plus in my view (thanks for working on this!).

Thus, as long as the JAX performance is equal (or even better), moving away from the additional NumPy seems like a good idea to me. For my understanding: Are there currently areas where NumPy+Numba is still noticeably faster (such as for small systems, for example)?

I do like PyTorch as a general ML framework and I find it easier to define general neural networks (starting already with ConvNets) in torch compared to stax.
However, we need to duplicate much of our sampling/optimization/SR logic for efficient use with PyTorch. This makes adding features cumbersome as you'd either (a) have to always implement everything for both frameworks (which may not always be possible, at least efficiently, and slows down development) or (b) having distinct feature sets in SR, time evolution, or similar for different machines (which would be quite unfortunate).
Also, the difference in interface between the functional-style JAX and the more object-oriented PyTorch further complicates supporting both at the same time.
Therefore, I agree with @PhilipVinc that removing PyTorch would help streamlining (and optimizing) our code.

It might be easier for people wanting to work with torch-based NQS to implement this outside of NetKet (ideally keeping some compatibility with regard to model definition, output formats, etc. to make it easy to interoperate and compare results), maybe starting with some of the code that's already there in our `v3.0` branch.

**Comment by tvieijra** [comment]
My two cents on this is that some types of layers (e.g. batchnorm with momentum) are pretty much impossible to implement in Stax, while they are implemented in more full-fledged neural network library such as PyTorch.  As @femtobit already pointed out, Stax in general is pretty minimal and sometimes hard to extend, as is also pointed out by JAX developers themselves.  Other, more complete libraries built on JAX, such as Trax can solve this issue.

**Comment by PhilipVinc** [comment]
Thanks for bringing this up.
However, I'd like to underline the fact that supporting jax means that we support anything that is built on top of it. 

If you define your neural network with any framework such as [Stax](https://jax.readthedocs.io/en/latest/), [Flax](https://github.com/google/flax), [Trax](https://github.com/google/trax), [Haiku](https://github.com/deepmind/dm-haiku), [Rlax](https://github.com/deepmind/rlax), all they do is reuse the [Jax](https://jax.readthedocs.io/en/latest/) primitives, so it's essentially a jax machine. If you then wrap it in our `JaxMachine` type. Everything should work™.

So the real question is not if Jax can stand against PyTorch, but rather if it's ecosystem can!

**Comment by tvieijra** [comment]
You are right that wrapping a neural network from any library as a `JaxMachine` is the way to go to use it with netket.  I think it might be worth it to provide examples on how to do this for the major libraries that you mentioned, if moving to JAX only.  I feel that wrapping a neural network from another library requires some knowledge about both netket and the libraries' internals, which some people might not want to acquire. 

The reason I brought this up was that we experienced problems wrapping a network as a `JaxMachine` with a network containing batchnorm layers.  One has to store the running average of the batchnorm layer somewhere, but this is impossible without JAX tracer issues.  We solved this by returning an additional variable containing these running averages, along with the parameters and the output state in the `apply_fn`.  However, this might be an edge case, as I am not really aware of other types of layers needing this functionality.  One other reason I could think of where this can be beneficial is when one wants to use a network which contains parameters that one does not want to update using it's gradient.  In any case, this thread might not be the place to discuss this issue.

**Comment by femtobit** [comment]
I think making it easy to directly use networks defined with Haiku (or one of the other JAX NN libraries, but Haiku seems pretty solid to me on first glance) would be the way to go.

Regarding the batch normalization issue @tvieijra mentioned: I think the solution of returning the additional state is the best approach for JAX (for example, [Haiku](https://github.com/deepmind/dm-haiku#working-with-non-trainable-state) does something very similar, also for batchnorm). This might be one of the things we could extend the `Jax` machine to support.

**Comment by PhilipVinc** [comment]
@tvieijra Please raise issues about this! If nobody speaks up we can't know about this sort of issues.

**Comment by PhilipVinc** [comment]
So, @gcarleo, do we go for this in 3.0 or do we release a 3.0 with deprecated non-jax bindings and then rather soonishly release a 3.1 where we remove all the deprecated things?

**Comment by fabienalet** [comment]
Sorry for late comment. As a mere practical user, it doesn't matter too much as both numpy and jax are easily available. However I would like to raise two points:
- it is quite important to keep an efficient mpi parallelization, which for me is a clear added value of Netket, to perform large-scale simulations. With Jax, this is done thanks to mpi+jax PR by @PhilipVinc , but will this be a long-term solution if jax changes?
- an efficient implementation of SR and symmetric machines is also an important advantage of Netket (again for large scale simulations), I didn't understand really what exactly will be lost if Numpy is evacuated? 

Related to @PhilipVinc 's last question, Overall I feel it might be too prematurate to completely remove Numpy in 3.0 (again as an average user, but of course the core developpers should decide)

**Comment by gcarleo** [comment]
Thanks @fabienalet, the goal is not to lose anything (i.e. have fast SR, MPI support, complex symmetric machines etc).
This can already be done with JAX and it is quite long term. The other advantage is that writing the derivativatives by hand for other applications (say unsupervised and others) is really cumbersome and time consuming when done in Numpy. Using Jax this is done automatically (see #482 ). The main reason why we cannot release 3.0 yet is that unsuperivsed and supervised learning are not implemented in numpy... 

My proposal @PhilipVinc @tvieijra @femtobit @fabienalet etc would be: 

1. merge #482 
2. quickly add supervised and unsupervised in jax only, using the feature in #482.
3. release 3.0 deprecating the numpy backend 
4. in 3.1 remove the numpy backend and reassess the situation for pytorch

**Comment by femtobit** [comment]
@gcarleo I very much agree with the general direction of moving to JAX as the primary supported backend. (By calling it *primary*, we can make it clear that this is the backend were we expect most development to happen and which users should use unless they have a strong requirement for PyTorch or similar, but leave the status of PyTorch open for the moment.)

Regarding the NumPy backend: I find it a bit strange to effectively introduce a new but already deprecated feature set in version 3.0. In understand this is primarily done to ease the transition from 2.x code to 3.0. However, if I am not mistaken we are already breaking the interface in 3.0 compared to 2.1 in several places (e.g., how symmetries are handled or the optimizers etc. taking the machine as argument). In that case, would it really be much more work to port 2.x code to 3.0 with JAX than to 3.0 with NumPy? There is some functionality missing at the moment (supervised/unsupervised comes to mind), but that is also true for plain NumPy in your proposal.

So, my question is: What features are currently missing from JAX (but present in NumPy) that would require someone switching from 2.1 to 3.0 to use the NumPy backend instead of moving straight to JAX? What I can think of is:
 * Lattice symmetries are supported by `RbmSpin` but not `JaxRbm`.
 * The `RbmSpin` has better performance than `JaxRbm` for some cases. (Is this still true?)
 * Maybe availability of JAX on some platforms? (Though I did not have issues with this so far.)
Are there other limitations like this?
